dora/src/masm/x64.rs

```diff
@@ -269,6 +269,41 @@ impl MacroAssembler {
         }
     }
 
+    pub fn float_cmp_total(
+        &mut self,
+        src_mode: MachineMode,
+        dest_mode: MachineMode,
+        dest: Reg,
+        lhs: FReg,
+        rhs: FReg,
+    ) {
+        self.float_as_int(dest_mode, RAX, src_mode, lhs);
+        self.float_as_int(dest_mode, RCX, src_mode, rhs);
+        self.float_as_int(dest_mode, RDX, src_mode, lhs);
+        self.float_as_int(dest_mode, RSI, src_mode, rhs);
+
+        self.load_int_const(dest_mode, REG_TMP1, 31);
+        self.int_sar(dest_mode, RDX, RDX, REG_TMP1);
+        self.load_int_const(dest_mode, REG_TMP1, 31);
+        self.int_sar(dest_mode, RSI, RSI, REG_TMP1);
+        self.load_int_const(dest_mode, REG_TMP2, 1);
+        self.int_shr(dest_mode, RDX, RDX, REG_TMP2);
+        self.int_shr(dest_mode, RSI, RSI, REG_TMP2);
+        self.int_xor(dest_mode, RDX, RDX, RAX);
+        self.int_xor(dest_mode, RSI, RSI, RCX);
+
+        self.int_xor(dest_mode, RCX, RCX, RCX);
+        self.cmp_reg(dest_mode, RDX, RSI);
+        self.set(RCX, CondCode::NotEqual);
+        self.load_int_const(dest_mode, RAX, -1);
+        let x64 = match dest_mode {
+            MachineMode::Int32 => 0,
+            MachineMode::Int64 => 1,
+            _ => unimplemented!(),
+        };
+        asm::cmov(self, x64, dest, RCX, CondCode::GreaterEq);
```

Comment:
This implementation is derived from https://godbolt.org/z/xgDe2q, which in turn is based on [__totalorderf](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/ieee754/flt-32/s_totalorderf.c;h=71ea5968ab3eadf575c42861f83a17092f803723;hb=0a8262a1b269f8f1602933248e69dfb041aab7a2):
```
        movd    eax, xmm0
        movd    ecx, xmm1
        mov     edx, eax
        sar     edx, 31
        mov     esi, ecx
        sar     esi, 31
        shr     edx
        xor     edx, eax
        shr     esi
        xor     esi, ecx
        xor     ecx, ecx
        cmp     edx, esi
        setne   cl
        mov     eax, -1
        cmovge  eax, ecx
        ret
```

It looks like I made some mistake translating it though, see the test file with comments.